packages/plugin-compiler/src/main/kotlin/io/realm/kotlin/compiler/IrUtils.kt

```diff
@@ -186,6 +195,30 @@ val ClassDescriptor.isEmbeddedRealmObject: Boolean
 val ClassDescriptor.isBaseRealmObject: Boolean
     get() = this.hasInterfacePsi(realmObjectPsiNames + embeddedRealmObjectPsiNames + asymmetricRealmObjectPsiNames) && !this.hasInterfacePsi(realmJavaObjectPsiNames)
 
+val realmObjectTypes = setOf(Name.identifier("RealmObject"), Name.identifier("EmbeddedRealmObject"), Name.identifier("AsymmetricRealmObject"))
+val realmObjectClassIds = realmObjectTypes.map { name -> ClassId(FqName("io.realm.kotlin.types"), name) }
+
+// This is the K2 equivalent of our PSI hack to determine if a symbol has a RealmObject base class.
+// There is currently no way to determine this within the resolved type system and there is
+// probably no such option around the corner.
+// https://kotlinlang.slack.com/archives/C03PK0PE257/p1694599154558669
+@OptIn(SymbolInternals::class)
+val FirClassSymbol<*>.isBaseRealmObject: Boolean
+    get() = this.classKind == ClassKind.CLASS &&
+        this.fir.superTypeRefs.any { typeRef ->
+            when (typeRef) {
+                // In SUPERTYPES stage
+                is FirUserTypeRef -> {
+                    typeRef.qualifier.last().name in realmObjectTypes &&
+                        // Disregard constructor invocations as that means that it is a Realm Java class
+                        !(typeRef.source?.run { treeStructure.getParent(lighterASTNode) }?.tokenType?.run { debugName == "CONSTRUCTOR_CALLEE" } ?: false)
```

Comment:
You can use `KtStubElementTypes.CONSTRUCTOR_CALLEE` instead of comparing the string.